File: parasmittal/Day 20/Design Circular Deque.cpp
class MyCircularDeque {
public:
    /** Initialize your data structure here. Set the size of the deque to be k. */
    vector<int> arr;
    int front,rear,size;
    MyCircularDeque(int k) {
        size=k;
        front=-1;
        rear=0;
        arr=vector<int>(size);
    }
    
    /** Adds an item at the front of Deque. Return true if the operation is successful. */
    bool insertFront(int value) {
        if(isFull())
            return false;
        if(front==-1){
            front=0;
            rear=0;
        }
        else if(front==0)
            front=size-1;
        else
            front=front-1;
        arr[front]=value;
        return true;
    }
    
    /** Adds an item at the rear of Deque. Return true if the operation is successful. */
    bool insertLast(int value) {
        if(isFull())
            return false;
        if(front==-1){
            front=0;
            rear=0;
        }
        else if(rear==size-1)
            rear=0;
        else
            rear=rear+1;
        arr[rear]=value;
        return true;
    }
    
    /** Deletes an item from the front of Deque. Return true if the operation is successful. */
    bool deleteFront() {
        if(isEmpty())
            return false;
        if(front==rear){
            front=-1;
            rear=-1;
        }
        else if(front==size-1)
            front=0;
        else
            front=front+1;
        return true;
    }
    
    /** Deletes an item from the rear of Deque. Return true if the operation is successful. */
    bool deleteLast() {
        if(isEmpty())
            return false;
        if(front==rear)
        {
            front=-1;
            rear=-1;
        }
        else if(rear==0)
            rear=size-1;
        else
            rear=rear-1;
        return true;
    }
    
    /** Get the front item from the deque. */
    int getFront() {
        if(isEmpty())
            return -1;
        return arr[front];
    }
    
    /** Get the last item from the deque. */
    int getRear() {
        if(isEmpty())
            return -1;
        return arr[rear];
    }
    
    /** Checks whether the circular deque is empty or not. */
    bool isEmpty() {
        if(front==-1)
            return true;
        return false;
        
    }
    
    /** Checks whether the circular deque is full or not. */
    bool isFull() {
        if((front==0 && rear==size-1) || front==rear+1)
            return true;
            return false;
        
    }
};

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque* obj = new MyCircularDeque(k);
 * bool param_1 = obj->insertFront(value);
 * bool param_2 = obj->insertLast(value);
 * bool param_3 = obj->deleteFront();
 * bool param_4 = obj->deleteLast();
 * int param_5 = obj->getFront();
 * int param_6 = obj->getRear();
 * bool param_7 = obj->isEmpty();
 * bool param_8 = obj->isFull();
 */

File: parnamondal/Week -4/Day-3/Design Circular Deque.py
# Time Complexity - O(1)
# Space Complexity - O(N)

class MyCircularDeque:

    def __init__(self, k: int):
        """
        Initialize your data structure here. Set the size of the deque to be k.
        """
        self.cq = []*k        
        self.size = 0
        self.max_len = k

    def insertFront(self, value: int) -> bool:
        """
        Adds an item at the front of Deque. Return true if the operation is successful.
        """
        if(self.size == self.max_len):
            return False
        self.cq.insert(0, value)
        self.size+=1
        return True

    def insertLast(self, value: int) -> bool:
        """
        Adds an item at the rear of Deque. Return true if the operation is successful.
        """
        if(self.size == self.max_len):
            return False
        self.cq.append(value)
        self.size+=1
        return True

    def deleteFront(self) -> bool:
        """
        Deletes an item from the front of Deque. Return true if the operation is successful.
        """
        if(not self.size):
            return False
        self.cq.pop(0)
        self.size-=1
        return True

    def deleteLast(self) -> bool:
        """
        Deletes an item from the rear of Deque. Return true if the operation is successful.
        """
        if(not self.size):
            return False
        self.cq.pop()
        self.size-=1
        return True

    def getFront(self) -> int:
        """
        Get the front item from the deque.
        """
        return -1 if not self.size else self.cq[0]

    def getRear(self) -> int:
        """
        Get the last item from the deque.
        """
        return -1 if not self.size else self.cq[-1]

    def isEmpty(self) -> bool:
        """
        Checks whether the circular deque is empty or not.
        """
        return not self.size

    def isFull(self) -> bool:
        """
        Checks whether the circular deque is full or not.
        """
        return self.size == self.max_len


# Your MyCircularDeque object will be instantiated and called as such:
# obj = MyCircularDeque(k)
# param_1 = obj.insertFront(value)
# param_2 = obj.insertLast(value)
# param_3 = obj.deleteFront()
# param_4 = obj.deleteLast()
# param_5 = obj.getFront()
# param_6 = obj.getRear()
# param_7 = obj.isEmpty()
# param_8 = obj.isFull()